### ISSUE TITLE
Interface as an object and union

### ISSUE THREAD

**Issue by Koxiaet** [async-graphql/async-graphql#241]
Support interfaces that are objects and a union, to avoid repeating the interface fields for each implementation. Example:

```rust
#[Interface]
struct Character {
    name: String,
    __implementations: CharacterTypes,
}

#[Union]
enum CharacterTypes {
    Human(Human),
    Droid(Droid),
}
```

Or, with fields:

```rust
#[Interface]
impl Shape {
    async fn area(&self) -> f32 { ... }
    async fn scale(&self, s: f32) -> Self { ... }
    async fn __implementations(&self) -> CharacterTypes { ... }
}
```
`__implementations` could also be called `__impls` because it's shorter.

**Comment by sunli829** [comment]
@Koxiaet @chipsenkbeil 

how about this? 🙂

```rust
#[derive(Interface)]
#[graphql(impl)]
enum CharacterTypes {
    Human(Human),
    Droid(Droid),
}

#[InterfaceImpl]
impl CharacterTypes {
    async fn id(&self) -> &str {
        todo!()
    }

    async fn name(&self) -> &str {
        todo!()
    }
}
```

**Comment by sunli829** [comment]
This cannot be verified at compile time. 🥺

> The object type must include a field of the same name for every field defined in an interface.

**Comment by Koxiaet** [comment]
The idea is that the outer struct provides all the common fields - the inner interface/union doesn't know about any fields of the outer one. It's for the case when the shared fields of the interface are resolved under common logic - so instead of being repeated for each variant in the interface, the code only has to be written once.

So ideally it would require no changes to the definition of the inner interface/union - the outer interface will simply add more fields to the inner one.

**Comment by sunli829** [comment]
Yes, so I'm still thinking about how to add fields to the objects inside.

**Comment by chipsenkbeil** [comment]
@sunli829 did you have any new thoughts about how to do this? I've got a trait that all of the variants in my union implement and it would be a lot easier to reference those trait methods like this

```graphql
query AllTheData {
    getSomeDataAsList {
        id
        type
        ... on Paragraph {
            text
        }
    }
}
```

instead of this

```graphql
query AllTheData {
    getSomeDataAsList {
        ... on Header {
            id
            type
        }
        ... on Section {
            id
            type
        }
        ... on Picture {
            id
            type
        }
        ... on Paragraph {
            id
            type
            text
        }
    }
}
```

**Comment by sunli829** [comment]
> @sunli829 did you have any new thoughts about how to do this? I could see it working where you implement the common methods on the enum, since that means you'd only need to write them once.
> 
> I've got a trait that all of the variants in my union implement and it would be a lot easier to reference those trait methods like this
> 
> ```graphql
> query AllTheData {
>     getSomeDataAsList {
>         id
>         type
>         ... on Paragraph {
>             text
>         }
>     }
> }
> ```
> 
> instead of this
> 
> ```graphql
> query AllTheData {
>     getSomeDataAsList {
>         ... on Header {
>             id
>             type
>         }
>         ... on Section {
>             id
>             type
>         }
>         ... on Picture {
>             id
>             type
>         }
>         ... on Paragraph {
>             id
>             type
>             text
>         }
>     }
> }
> ```

I haven't figured out how to achieve this.

**Comment by chipsenkbeil** [comment]
@sunli829, I was skimming through some of the [ComplexObject macro code](https://github.com/async-graphql/async-graphql/commit/bec0fc18a98a0385b3a46795c715cc673fde15ef). Could something similar be done for a union with a "ComplexUnion" feature?

**Comment by sunli829** [comment]
> @sunli829, I was skimming through some of the [ComplexObject macro code](https://github.com/async-graphql/async-graphql/commit/bec0fc18a98a0385b3a46795c715cc673fde15ef). Could something similar be done for a union with a "ComplexUnion" feature?

How to use this `ComplexUnion`?

**Comment by chipsenkbeil** [comment]
@sunli829 from a library user's perspective, something like this:

```rust
use async_graphql::*;

#[derive(Union)]
#[graphql(complex)] // NOTE: If you want the `ComplexObject` macro to take effect, this `complex` attribute is required.
enum Shape {
    Circle(Circle),
    Square(Square),
}


#[ComplexUnion]
impl Shape {
    async fn area(&self) -> f32 {
        match self {
            Self::Circle(x) => std::f32::consts::PI * x.radius * x.radius,
            Self::Square(x) => x.width * x.width,
        }
        
    }

    async fn scale(&self, s: f32) -> Shape {
        match self {
            Self::Circle(x) => Circle { radius: x.radius * s }.into(),
            Self::Square(x) => Square { width: x.width * s }.into()
        }
    }
}

struct Circle {
    radius: f32,
}

#[Object]
impl Circle {
    async fn radius(&self) -> f32 {
        self.radius
    }
}

struct Square {
    width: f32,
}

#[Object]
impl Square {
    async fn width(&self) -> f32 {
        self.width
    }
}
```

enables a graphql query like

```graphql
{
    some_shape(...) {
        area
        scale {
            area
        }
        ... on Circle {
            radius
        }
        ... on Square {
            width
        }
    }
}
```

This follows the interface of graphql (instead of union), but enables implementing all of the common fields in a single place instead of needing to repeat them for every single object:

<img width="776" alt="image" src="https://user-images.githubusercontent.com/2481802/121120566-6de28f00-c7e3-11eb-9fa9-064c25533031.png">

**Comment by bugproof** [comment]
I see it can be a little bit tricky to think of a good way to solve this because rust doesn't support subtypes/inheritance to inherit fields, only composition. So this can be solved only with some macro magic.

**Comment by mrzahrada** [comment]
Hi, is there any progress on this issue? thx :)

**Comment by PookieBuns** [comment]
> @Koxiaet @chipsenkbeil
> 
> how about this? 🙂
> 
> ```rust
> #[derive(Interface)]
> #[graphql(impl)]
> enum CharacterTypes {
>     Human(Human),
>     Droid(Droid),
> }
> 
> #[InterfaceImpl]
> impl CharacterTypes {
>     async fn id(&self) -> &str {
>         todo!()
>     }
> 
>     async fn name(&self) -> &str {
>         todo!()
>     }
> }
> ```

Is there any plan on supporting this?
